File: companion-ios/Sources/FITDecoder.swift
import Foundation
import CoreLocation

// A parsed ride, ready to preview.
struct RidePreview {
    struct Point {
        let coordinate: CLLocationCoordinate2D
        let altitude: Double?
        let speedKmh: Double?
        let power: Int?
        let heartRate: Int?
        let cadence: Int?
    }
    var points: [Point] = []
    var start: Date?
    var duration: TimeInterval = 0
    var distanceKm: Double = 0
    var avgSpeedKmh: Double = 0
    var maxSpeedKmh: Double = 0
    var avgPower: Int?
    var avgHeartRate: Int?
    var ascentM: Double = 0

    var coordinates: [CLLocationCoordinate2D] { points.map(\.coordinate) }
}

// Minimal FIT decoder for the files this device writes (see
// firmware src/fit_writer.cpp). It reads definition + data messages
// generically and interprets the record message (global 20).
enum FITDecoder {
    private static let epochOffset: TimeInterval = 631_065_600  // 1989-12-31
    private static let semicircle = 180.0 / 2147483648.0

    struct FieldDef { let num: Int; let size: Int }
    struct MsgDef { let global: Int; let fields: [FieldDef] }

    static func decode(_ data: Data) -> RidePreview? {
        let b = [UInt8](data)
        guard b.count > 14 else { return nil }

        let headerSize = Int(b[0])
        guard b.count >= headerSize else { return nil }
        let dataSize = Int(b[4]) | (Int(b[5]) << 8) | (Int(b[6]) << 16) | (Int(b[7]) << 24)
        var i = headerSize
        // A ride that lost power mid-recording never got its header patched
        // (data_size = 0) or trailer/CRC written. In that case scan to the end
        // of the file — the record messages before the cut are still valid, and
        // the bounds/undefined-type checks below stop cleanly at any trailing
        // partial write or padding.
        var end = headerSize + dataSize
        if dataSize == 0 || end > b.count || end < headerSize {
            end = b.count
        }

        var defs: [Int: MsgDef] = [:]
        var preview = RidePreview()
        var firstTime: UInt32? = nil
        var lastTime: UInt32 = 0
        var powerSum = 0, powerN = 0, hrSum = 0, hrN = 0
        var climbBase: Double? = nil   // 3 m-hysteresis ascent, like the device

        while i < end {
            let rec = b[i]; i += 1
            if rec & 0x40 != 0 {                    // definition message
                let local = Int(rec & 0x0F)
                guard i + 5 <= end else { break }
                // b[i]=reserved, b[i+1]=arch (0=LE), b[i+2..3]=global, b[i+4]=count
                let global = Int(b[i + 2]) | (Int(b[i + 3]) << 8)
                let count = Int(b[i + 4])
                i += 5
                var fields: [FieldDef] = []
                for _ in 0..<count {
                    guard i + 3 <= end else { break }
                    fields.append(FieldDef(num: Int(b[i]), size: Int(b[i + 1])))
                    i += 3
                }
                // Developer fields (0x20) — this device never emits them.
                defs[local] = MsgDef(global: global, fields: fields)
            } else {                                // data message
                let local = Int(rec & 0x0F)
                guard let def = defs[local] else { break }
                var off = i
                var lat: Int32? = nil, lon: Int32? = nil
                var alt: Double? = nil, speed: Double? = nil
                var power: Int? = nil, hr: Int? = nil, cad: Int? = nil
                for f in def.fields {
                    guard off + f.size <= end else { break }
                    if def.global == 20 {           // record
                        switch f.num {
                        case 253: lastTime = u32(b, off)
                                  if firstTime == nil { firstTime = lastTime }
                        case 0:   let v = Int32(bitPattern: u32(b, off))
                                  if v != Int32(bitPattern: 0x7FFFFFFF) { lat = v }
                        case 1:   let v = Int32(bitPattern: u32(b, off))
                                  if v != Int32(bitPattern: 0x7FFFFFFF) { lon = v }
                        case 2:   let v = u16(b, off)
                                  if v != 0xFFFF { alt = Double(v) / 5.0 - 500.0 }
                        case 6:   let v = u16(b, off)
                                  if v != 0xFFFF { speed = Double(v) / 1000.0 * 3.6 }
                        case 7:   let v = u16(b, off); if v != 0xFFFF { power = Int(v) }
                        case 3:   let v = b[off]; if v != 0xFF { hr = Int(v) }
                        case 4:   let v = b[off]; if v != 0xFF { cad = Int(v) }
                        default: break
                        }
                    }
                    off += f.size
                }
                i = off
                if def.global == 20, let lat, let lon {
                    let coord = CLLocationCoordinate2D(
                        latitude: Double(lat) * semicircle,
                        longitude: Double(lon) * semicircle)
                    preview.points.append(.init(
                        coordinate: coord, altitude: alt, speedKmh: speed,
                        power: power, heartRate: hr, cadence: cad))
                    if let s = speed { preview.maxSpeedKmh = max(preview.maxSpeedKmh, s) }
                    if let p = power { powerSum += p; powerN += 1 }
                    if let h = hr { hrSum += h; hrN += 1 }
                    if let a = alt {
                        if let base = climbBase {
                            if a > base + 3 { preview.ascentM += a - base; climbBase = a }
                            else if a < base - 3 { climbBase = a }
                        } else {
                            climbBase = a
                        }
                    }
                }
            }
        }

        guard preview.points.count >= 2 else { return nil }

        if let ft = firstTime {
            preview.start = Date(timeIntervalSince1970: epochOffset + Double(ft))
            preview.duration = Double(lastTime - ft)
        }
        preview.distanceKm = trackDistanceKm(preview.coordinates)
        if preview.duration > 0 {
            preview.avgSpeedKmh = preview.distanceKm / (preview.duration / 3600)
        }
        preview.avgPower = powerN > 0 ? powerSum / powerN : nil
        preview.avgHeartRate = hrN > 0 ? hrSum / hrN : nil
        return preview
    }

    private static func u16(_ b: [UInt8], _ o: Int) -> UInt16 {
        UInt16(b[o]) | (UInt16(b[o + 1]) << 8)
    }
    private static func u32(_ b: [UInt8], _ o: Int) -> UInt32 {
        UInt32(b[o]) | (UInt32(b[o + 1]) << 8) | (UInt32(b[o + 2]) << 16) | (UInt32(b[o + 3]) << 24)
    }

    private static func trackDistanceKm(_ c: [CLLocationCoordinate2D]) -> Double {
        guard c.count > 1 else { return 0 }
        var m = 0.0
        for i in 1..<c.count {
            m += CLLocation(latitude: c[i-1].latitude, longitude: c[i-1].longitude)
                .distance(from: CLLocation(latitude: c[i].latitude, longitude: c[i].longitude))
        }
        return m / 1000
    }
}

File: companion-android/app/src/main/java/com/raemond/opentrailpaper/data/FitDecoder.kt
package com.raemond.opentrailpaper.data

import kotlin.math.max

/** A parsed ride, ready to preview. */
data class RidePreview(
    val points: List<Point> = emptyList(),
    val startEpochMs: Long? = null,
    val duration: Double = 0.0,      // elapsed wall-clock, seconds
    val movingTime: Double = 0.0,    // device "MOVING TIME" (session timer)
    val distanceKm: Double = 0.0,
    val avgSpeedKmh: Double = 0.0,
    val maxSpeedKmh: Double = 0.0,
    val avgPower: Int? = null,
    val normPower: Int? = null,
    val avgHeartRate: Int? = null,
    val ascentM: Double = 0.0,
) {
    data class Point(
        val coordinate: LatLon,
        val altitude: Double?,
        val speedKmh: Double?,
        val power: Int?,
        val heartRate: Int?,
        val cadence: Int?,
    )

    val coordinates: List<LatLon> get() = points.map { it.coordinate }
}

/**
 * Minimal FIT decoder for the files this device writes (see firmware
 * src/fit_writer.cpp). It reads definition + data messages generically and
 * interprets the record message (global 20) and the session roll-up (global 18).
 *
 * A port of companion-ios/Sources/FITDecoder.swift, field-for-field.
 */
object FitDecoder {
    private const val EPOCH_OFFSET_SEC = 631_065_600.0   // 1989-12-31
    private const val SEMICIRCLE = 180.0 / 2147483648.0

    private data class FieldDef(val num: Int, val size: Int)
    private data class MsgDef(val global: Int, val fields: List<FieldDef>)

    fun decode(data: ByteArray): RidePreview? {
        val b = data
        if (b.size <= 14) return null

        val headerSize = b[0].toInt() and 0xFF
        if (b.size < headerSize) return null
        val dataSize = le32(b, 4)
        var i = headerSize
        // A ride that lost power mid-recording never got its header patched
        // (data_size = 0) or trailer/CRC written. In that case scan to the end of
        // the file — the record messages before the cut are still valid, and the
        // bounds checks below stop cleanly at any trailing partial write.
        var end = headerSize + dataSize
        if (dataSize == 0 || end > b.size || end < headerSize) end = b.size

        val defs = HashMap<Int, MsgDef>()
        val points = ArrayList<RidePreview.Point>()
        var firstTime: Long? = null
        var lastTime = 0L
        var powerSum = 0; var powerN = 0
        var hrSum = 0; var hrN = 0
        var maxSpeed = 0.0
        var ascent = 0.0
        var climbBase: Double? = null   // 3 m-hysteresis ascent, like the device

        // Session roll-up (global 18) — the numbers the device shows on its
        // ride-complete screen. Preferred over the record-derived values below
        // when present (older files without a summary fall back to derived).
        var sMoving: Double? = null
        var sDistKm: Double? = null
        var sAvgSpeed: Double? = null
        var sAvgPower: Int? = null
        var sNormPower: Int? = null
        var sAvgHr: Int? = null
        var sAscent: Double? = null

        while (i < end) {
            val rec = b[i].toInt() and 0xFF
            i += 1
            if (rec and 0x40 != 0) {                          // definition message
                val local = rec and 0x0F
                if (i + 5 > end) break
                // b[i]=reserved, b[i+1]=arch (0=LE), b[i+2..3]=global, b[i+4]=count
                val global = (b[i + 2].toInt() and 0xFF) or ((b[i + 3].toInt() and 0xFF) shl 8)
                val count = b[i + 4].toInt() and 0xFF
                i += 5
                val fields = ArrayList<FieldDef>(count)
                for (f in 0 until count) {
                    if (i + 3 > end) break
                    fields.add(FieldDef(b[i].toInt() and 0xFF, b[i + 1].toInt() and 0xFF))
                    i += 3
                }
                // Developer fields (0x20) — this device never emits them.
                defs[local] = MsgDef(global, fields)
            } else {                                          // data message
                val def = defs[rec and 0x0F] ?: break
                var off = i
                var lat: Int? = null
                var lon: Int? = null
                var alt: Double? = null
                var speed: Double? = null
                var power: Int? = null
                var hr: Int? = null
                var cad: Int? = null
                for (f in def.fields) {
                    if (off + f.size > end) break
                    if (def.global == 20) {                   // record
                        when (f.num) {
                            253 -> {
                                lastTime = le32(b, off).toLong() and 0xFFFFFFFFL
                                if (firstTime == null) firstTime = lastTime
                            }

                            0 -> le32(b, off).let { if (it != 0x7FFFFFFF) lat = it }
                            1 -> le32(b, off).let { if (it != 0x7FFFFFFF) lon = it }
                            2 -> le16(b, off).let { if (it != 0xFFFF) alt = it / 5.0 - 500.0 }
                            6 -> le16(b, off).let { if (it != 0xFFFF) speed = it / 1000.0 * 3.6 }
                            7 -> le16(b, off).let { if (it != 0xFFFF) power = it }
                            3 -> (b[off].toInt() and 0xFF).let { if (it != 0xFF) hr = it }
                            4 -> (b[off].toInt() and 0xFF).let { if (it != 0xFF) cad = it }
                        }
                    } else if (def.global == 18) {            // session roll-up
                        when (f.num) {
                            8 -> sMoving = (le32(b, off).toLong() and 0xFFFFFFFFL) / 1000.0
                            9 -> sDistKm = (le32(b, off).toLong() and 0xFFFFFFFFL) / 100.0 / 1000.0
                            14 -> le16(b, off).let { if (it != 0xFFFF) sAvgSpeed = it / 1000.0 * 3.6 }
                            16 -> (b[off].toInt() and 0xFF).let { if (it != 0xFF) sAvgHr = it }
                            20 -> le16(b, off).let { if (it != 0xFFFF) sAvgPower = it }
                            34 -> le16(b, off).let { if (it != 0xFFFF) sNormPower = it }
                            22 -> le16(b, off).let { if (it != 0xFFFF) sAscent = it.toDouble() }
                        }
                    }
                    off += f.size
                }
                i = off
                val la = lat
                val lo = lon
                if (def.global == 20 && la != null && lo != null) {
                    points.add(
                        RidePreview.Point(
                            coordinate = LatLon(la * SEMICIRCLE, lo * SEMICIRCLE),
                            altitude = alt,
                            speedKmh = speed,
                            power = power,
                            heartRate = hr,
                            cadence = cad,
                        ),
                    )
                    speed?.let { maxSpeed = max(maxSpeed, it) }
                    power?.let { powerSum += it; powerN += 1 }
                    hr?.let { hrSum += it; hrN += 1 }
                    alt?.let { a ->
                        val base = climbBase
                        if (base == null) {
                            climbBase = a
                        } else if (a > base + 3) {
                            ascent += a - base; climbBase = a
                        } else if (a < base - 3) {
                            climbBase = a
                        }
                    }
                }
            }
        }

        if (points.size < 2) return null

        val ft = firstTime
        val startMs = ft?.let { ((EPOCH_OFFSET_SEC + it) * 1000).toLong() }
        val duration = if (ft != null) (lastTime - ft).toDouble() else 0.0
        var distanceKm = trackDistanceKm(points.map { it.coordinate })
        var avgSpeed = if (duration > 0) distanceKm / (duration / 3600) else 0.0
        var avgPower = if (powerN > 0) powerSum / powerN else null
        var avgHr = if (hrN > 0) hrSum / hrN else null

        // Prefer the device's session roll-up so the app matches the
        // ride-complete screen exactly (incl. the map-DEM ascent, which can't be
        // re-derived).
        sDistKm?.let { if (it > 0) distanceKm = it }
        sAvgSpeed?.let { avgSpeed = it }
        sAvgPower?.let { avgPower = it }
        sAvgHr?.let { avgHr = it }

        return RidePreview(
            points = points,
            startEpochMs = startMs,
            duration = duration,
            movingTime = sMoving ?: duration,
            distanceKm = distanceKm,
            avgSpeedKmh = avgSpeed,
            maxSpeedKmh = maxSpeed,
            avgPower = avgPower,
            normPower = sNormPower,
            avgHeartRate = avgHr,
            ascentM = sAscent ?: ascent,
        )
    }

    private fun le16(b: ByteArray, o: Int): Int =
        (b[o].toInt() and 0xFF) or ((b[o + 1].toInt() and 0xFF) shl 8)

    private fun le32(b: ByteArray, o: Int): Int =
        (b[o].toInt() and 0xFF) or ((b[o + 1].toInt() and 0xFF) shl 8) or
            ((b[o + 2].toInt() and 0xFF) shl 16) or ((b[o + 3].toInt() and 0xFF) shl 24)

    private fun trackDistanceKm(c: List<LatLon>): Double {
        if (c.size < 2) return 0.0
        var m = 0.0
        for (i in 1 until c.size) m += c[i - 1].distanceTo(c[i])
        return m / 1000
    }
}
